### PR TITLE
feat(pet-7): frequency tracking + escalation tiers

### PR DESCRIPTION
**Pull request by ziomancer**
## Summary
- **FrequencyTracker**: per-session exponential decay scoring with rolling window promotion, LRU eviction (prefer terminated sessions), rate limiting for new session creation, and configurable weight matching (exact > glob by longest prefix)
- **Escalation module**: 3-tier evaluation (Tier 1: deep_inspect, Tier 2: enhanced_scrutiny, Tier 3: terminate) with hardcoded Tier 3 floor of 30.0 that cannot be disabled
- **Pipeline integration**: `activate()`/`deactivate()` instance methods, double-gated frequency + escalation hooks at pipeline stages 6-7, `_build_result()` helper for consistent PipelineResult construction, `premium_features` manifest (MappingProxyType)

## Layered defense

The frequency tracker and escalation module compose as two stages in the pipeline:

1. **Frequency hook** (stage 6): computes per-session score via exponential decay + weighted rule matches. Rolling window counts findings within a time window and promotes the tier if the threshold is exceeded.
2. **Escalation hook** (stage 7): reads the frequency result and maps the score to a tier using configurable thresholds (>= comparison, matching Drawbridge). Tier 3 floor is hardcoded at 30.0 — config validation rejects any attempt to set it lower.

Both hooks are double-gated: `_check_premium()` (global premium active) AND per-feature config toggle (`frequency_enabled`, `escalation_enabled`). When premium is inactive, hooks are no-ops and PipelineResult fields default to None.

`safe` is independent of escalation tier — Tier 3 does not force `safe=False` (D11).

## Files changed

| File | Change |
|------|--------|
| `petasos/premium/frequency.py` | New — FrequencyTracker, SessionState, FrequencyUpdateResult, weight matching, eviction, rate limiting |
| `petasos/premium/escalation.py` | New — evaluate_tier(), evaluate_escalation(), EscalationResult, TIER3_FLOOR |
| `petasos/premium/__init__.py` | New — premium package exports |
| `petasos/pipeline.py` | Adds activate/deactivate, _build_result, frequency + escalation hooks |
| `petasos/_types.py` | Adds escalation_tier, session_score, premium_features to PipelineResult |
| `petasos/config.py` | Adds 10 premium config fields + validation (positive, finite, ascending, floor) |
| `petasos/__init__.py` | Exports FrequencyTracker, FrequencyUpdateResult |
| `tests/test_frequency.py` | New — 29 tests (decay math, weights, rolling window, eviction, rate limiting, edge cases) |
| `tests/test_escalation.py` | New — 15 tests (tier eval, floor, escalation result, frozen) |
| `tests/test_premium_integration.py` | New — 19 tests (pipeline hooks, result fields, config validation, activate/deactivate, exception handler) |

## Test plan

- [x] `python -m pytest tests/test_frequency.py tests/test_escalation.py tests/test_premium_integration.py tests/test_pipeline.py tests/test_config.py -v` — 124/124 pass (full output: docs/specs/TODO/PET-7.test-output.txt)
- [x] `python -m mypy --strict petasos/premium/frequency.py petasos/premium/escalation.py petasos/pipeline.py petasos/config.py petasos/_types.py` — clean
- [x] `python -m ruff check petasos/premium/ petasos/pipeline.py petasos/config.py petasos/_types.py` — clean

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Premium frequency tracking: per-session scoring with decay, rolling windows, rate limits, promotions, and termination
  * Tier-based escalation: configurable thresholds, actions, and top-tier termination
  * Pipeline premium toggle: activate/deactivate premium behavior; results now include session score, escalation tier, and a premium feature manifest
  * New premium configuration options with validation for thresholds, weights, sessions, and rate limits

* **Documentation**
  * Captured test run output added to docs

* **Tests**
  * Extensive unit and integration tests covering frequency, escalation, config validation, and pipeline premium behavior

[Review Change Stack](https://app.coderabbit.ai/change-stack/Vigil-Harbor/Petasos/pull/6?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: petasos/config.py [hunk 1]
```diff
@@ -26,14 +29,30 @@ class PetasosConfig:
...
     frequency_enabled: bool = False
     escalation_enabled: bool = False
     profile_name: str | None = None
     tool_guard_enabled: bool = False
     audit_enabled: bool = False
     alert_enabled: bool = False
 
+    # Frequency tracking
+    frequency_half_life_seconds: float = 60.0
+    frequency_weights: dict[str, float] | None = None
```
[the remark this answers is not in the document]
Comment (ziomancer) [reply petasos/config.py]: Resolved in 456c66e

File: petasos/config.py [hunk 2]
```diff
@@ -54,6 +73,65 @@ def __post_init__(self) -> None:
...
+                f"got {self.max_new_sessions_per_minute!r}"
+            )
+        if self.frequency_weights is not None:
+            for k, v in self.frequency_weights.items():
+                if v < 0 or not math.isfinite(v):
+                    raise ValueError(
+                        f"frequency_weights values must be non-negative and finite, "
+                        f"got {k!r}: {v!r}"
+                    )
+            object.__setattr__(self, "frequency_weights", dict(self.frequency_weights))
```
[the remark this answers is not in the document]
Comment (ziomancer) [reply petasos/config.py]: Resolved in 456c66e

File: petasos/premium/frequency.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        # Step 11: Return
+        return FrequencyUpdateResult(
+            previous_score=previous_score,
+            current_score=current_score,
+            tier=tier,
+            terminated=state.terminated,
+        )
+
+    def get_state(self, session_id: str) -> SessionState | None:
+        return self._sessions.get(session_id)
```
[the remark this answers is not in the document]
Comment (ziomancer) [reply petasos/premium/frequency.py]: Resolved in 456c66e

File: tests/test_escalation.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
+from __future__ import annotations
+
+import pytest
+
+from petasos.config import PetasosConfig
+from petasos.premium.escalation import (
+    TIER3_FLOOR,
+    EscalationResult,
```
[the remark this answers is not in the document]
Comment (ziomancer) [reply tests/test_escalation.py]: Resolved in 456c66e

File: tests/test_frequency.py [hunk 5]
```diff
@@ -0,0 +1,491 @@
+from __future__ import annotations
+
+import math
+import time
```
[the remark this answers is not in the document]
Comment (ziomancer) [reply tests/test_frequency.py]: Resolved in 456c66e

File: tests/test_premium_integration.py [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+
+import asyncio
+from types import MappingProxyType
+from unittest.mock import patch
+
+import pytest
+
+from petasos._types import PipelineResult, ScanFinding, Severity, Position
+from petasos.config import PetasosConfig
+from petasos.pipeline import Pipeline
```
[the remark this answers is not in the document]
Comment (ziomancer) [reply tests/test_premium_integration.py]: Resolved in 456c66e

File: tests/test_premium_integration.py [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+# ---------------------------------------------------------------------------
+
+
+class TestPipelineHooks:
+    def test_premium_inactive_hooks_are_noop(self) -> None:
+        cfg = _cfg()
+        pipe = Pipeline(config=cfg)
+        result = asyncio.get_event_loop().run_until_complete(
+            pipe.inspect("hello", session_id="s1")
+        )
```
[the remark this answers is not in the document]
Comment (ziomancer) [reply tests/test_premium_integration.py]: Resolved in 456c66e

File: petasos/__init__.py [hunk 8]
```diff
@@ -11,10 +11,13 @@
 from petasos.config import PetasosConfig
 from petasos.normalize import normalize
 from petasos.pipeline import Pipeline
+from petasos.premium.frequency import FrequencyTracker, FrequencyUpdateResult
```
[the remark this answers is not in the document]
Comment (ziomancer) [reply petasos/__init__.py:14]: Skipped — PetasosConfig is an intentional public export; re-exporting from __init__.py is standard Python packaging practice for the API surface.

File: petasos/config.py [hunk 9]
```diff
@@ -54,12 +73,79 @@ def __post_init__(self) -> None:
...
+                        f"frequency_weights keys must be non-empty strings, got {k!r}"
+                    )
+                if v < 0 or not math.isfinite(v):
+                    raise ValueError(
+                        f"frequency_weights values must be non-negative and finite, "
+                        f"got {k!r}: {v!r}"
+                    )
+            object.__setattr__(
+                self, "frequency_weights", MappingProxyType(dict(self.frequency_weights))
+            )
```
[the remark this answers is not in the document]
Comment (ziomancer) [reply petasos/config.py:140]: Resolved in 0186e81 — ruff format applied to all files.

File: tests/test_premium_integration.py [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+    return PetasosConfig(**defaults)  # type: ignore[arg-type]
+
+
+# ---------------------------------------------------------------------------
+# Pipeline hooks
+# ---------------------------------------------------------------------------
+
+
+class TestPipelineHooks:
+    async def test_premium_inactive_hooks_are_noop(self) -> None:
```
[the remark this answers is not in the document]
Comment (ziomancer) [reply tests/test_premium_integration.py:28]: Skipped — project uses pytest-asyncio with asyncio_mode="auto" (pyproject.toml), not anyio. All async tests run correctly without markers. Adding @pytest.mark.anyio would be incorrect for this project.
